### PR TITLE
fsutil: Add SlashedPathDir() and SlashedPathClean() functions

### PR DESCRIPTION
**Pull request by woky**
From the code:

  These functions exists because we work with slash terminated
  directory paths that come from deb package tarballs but standard
  library path functions treat slash terminated paths as unclean.

We use ad-hoc code to make filepath.Dir() slash terminated in two places
right now. For example this code

  targetDir := filepath.Dir(strings.TrimRight(targetPath, "/")) + "/"

is not strictly correct as "/a/b/.///" will be turned into "/a/b/./"
which is still "/a/b".

Since we deal with slash terminated directory paths everywhere, create
and use dedicated helper functions to process them.

### CODE REVIEW

**Review by rebornplusplus — APPROVED** [review]
Nice.

File: internal/fsutil/path.go
Comment (Aflynn50) [inline internal/fsutil/path.go:10]: This text seems like it would belong better in the PR description than in the code. For me at least, I would expect the documentation for the `Dir` function to just contain information about what exactly it does, rather than information about why it is needed. If someone is wondering why the code has used `fsutil.Dir` instead of `filepath.Dir` they could look at the code here and see that it does some extra cleaning.

File: internal/fsutil/path.go
Comment (Aflynn50) [inline internal/fsutil/path.go:13]: Function comments should start with the name of the function e.g. `// Dir cleans the filepath and then returns all but the last element of the path. It includes a trailing slash.`

File: internal/fsutil/path.go
Comment (Aflynn50) [inline internal/fsutil/path.go:21]: Again, this comment makes sense in the PR description, but not really in the code. An example of what this `Dir` function returns would be good, but I don't personally see much reason to have the comparison. When someone wants to use this function they just need to know what its going to return, its reason for existing should be obvious from that (i.e. they can look at the code and see that `filepath.Dir` is called inside the function).

File: internal/fsutil/path.go
Comment (Aflynn50) [inline internal/fsutil/path.go:40]: Same points as for the comments on Dir.

File: internal/fsutil/path.go
Comment (Aflynn50) [inline internal/fsutil/path.go:30]: Rather than directing people to documentation for `filepath.Clean()` it could be better to write exactly what it does here. (they may not necceserily know exactly what `filepath.Clean` does.)

File: internal/fsutil/path_test.go
Comment (Aflynn50) [inline internal/fsutil/path_test.go:9]: Could `dirTests` and `cleanTests` be declared inside the test rather than globally?

File: internal/slicer/slicer.go
Comment (Aflynn50) [inline internal/slicer/slicer.go:39]: Could the path variable from before it is cleaned be called something like rawPath to make it clear that it has not been cleaned? And then you can refer to it simply as path from thereon out.

**Review by Aflynn50 — COMMENTED** [review]
Looking good, just a few pedantic things.

**Review by cjdcordeiro — APPROVED** [review]
Thanks! @Aflynn50 comments make sense to me ;)

File: internal/slicer/slicer.go
Comment (flotter) [inline internal/slicer/slicer.go:39]: The output of these two snippets are not identical I believe?

```slashpath``` will always end with a ```Separator```.
```path``` uses ```fsutil.Clean()``` only, which could end without a ```Separator```.

File: internal/slicer/slicer.go
Comment (flotter) [inline internal/slicer/slicer.go:40]: I really think this loop can use some comments, taking us through the logic. For example, the ```filepath.Clean()``` is already iterative, so my first question is why do we need the loop.

**Review by flotter — COMMENTED** [review]
This is a great initiative for simplifying the code. Some minor tweak proposals.

File: internal/fsutil/path.go
Comment (woky) [reply internal/fsutil/path.go:10]: This is subjective. I like the the function comment with examples more. When someone who is familiar with `filepath.Dir()` sees usages of `fsutil.Dir()` the first question they will ask is _Why don't they just use `filepath.Dir`?_ and this comment gives that answer. I don't think the documentation with examples hurts anyone. It's not that verbose to get in the way of reading the code.

File: internal/fsutil/path.go
Comment (woky) [reply internal/fsutil/path.go:13]: Updated the comment to start with function name, but kept the description in terms of `filepath.Dir()` so that it's obvious why it exists and how is it different.

File: internal/fsutil/path.go
Comment (woky) [reply internal/fsutil/path.go:21]: Replied above. Resolving this one.

File: internal/fsutil/path.go
Comment (woky) [reply internal/fsutil/path.go:40]: Replied above. Resolving this one.

File: internal/fsutil/path.go
Comment (woky) [reply internal/fsutil/path.go:30]: Replied above. Resolving this one.

File: internal/fsutil/path_test.go
Comment (woky) [reply internal/fsutil/path_test.go:9]: This would deviate from how it's currently done in Chisel since its inception before I even started contributing. I'll let the maintainer decide on this one.

File: internal/slicer/slicer.go
Comment (woky) [reply internal/slicer/slicer.go:39]: Done.

File: internal/fsutil/path.go
Comment (woky) [reply internal/fsutil/path.go:1]: > Also, can you help me understand the value of having both Clean and Dir ? For example, why cannot there be a single function that clean and slash the dir, as in why do we need a separate clean (since Dir() also cleans in any case)?

This is the consequence of the PR rules. You can see the full diff being used in followup PR https://github.com/canonical/chisel/pull/77/files where both `fsutil.Dir()` and `fsutil.Clean()` are used. The latter is used in this commit.

For more context, please see this comment from Gustavo: https://github.com/canonical/chisel/pull/44#pullrequestreview-1242596246

> I'd like to understand what the issue is in more detail, because this change seems to be breaking down an assumption that has always been present in the code, and that we should attempt not to break: directories end up in slashes, while files don't. We can see that throughout the code, and through the testing as well. Most of the changes in this PR apparently go towards making this less consistent, which feels like the opposite direction we should be going.

With regards to naming, I don't mind changing it. What would you call `fsutil.Clean()` which only appends slash for regular file paths?

File: internal/fsutil/path.go
Comment (woky) [reply internal/fsutil/path.go:38]: That would turn regular file path `/foo/bar` into directory path `/foo/bar/`. Please see the usage of `fsutil.Clean()` in this commit. If I were to only use it for directory paths, it'd defeat the purpose of abstracting away the convention of ending every directory path with slash. I.e. instead of
```
path := fsutil.Clean()
```
now I'd have
```
var path string
if strings.HasSuffix("/") {
  path = fsutil.Clean(path)
} else {
  path = filepath.Clean(path)
}
```
which doesn't seem like an improvement over existing code. See the example in the commit message where we do this wrong in one place and Gustavo's comment I linked above about the path format assumption.

File: internal/slicer/slicer.go
Comment (woky) [reply internal/slicer/slicer.go:39]: Can you elaborate on why you think `slashpath` will always end with separator? From diff snippet it looks like it'll only end with separator when path ends with `/`'.

File: internal/slicer/slicer.go
Comment (woky) [reply internal/slicer/slicer.go:40]: I've added the comment to the (lambda) function:
```
	// addKnownPath path adds path and all its directory parent paths into
	// knownPaths set.
```
This code was introduced the PR from which I linked Gustavo's comment. It's used for checking which paths can be listed by scriptlets.

File: internal/fsutil/path.go
Comment (Aflynn50) [reply internal/fsutil/path.go:10]: In my opinion the comments above the functions themselves are enough for this, to me this only serves as an extra block of text to parse and comprehend that doesn't aid my overall understanding

File: internal/fsutil/path.go
Comment (flotter) [reply internal/fsutil/path.go:38]: I missed the fact that both files and directories needs to be handled. Thanks for explaining this to me.

File: internal/fsutil/path_test.go
Comment (flotter) [inline internal/fsutil/path_test.go:32]: I think you should add "/" root here :)

File: internal/slicer/slicer.go
Comment (woky) [reply internal/slicer/slicer.go]: Thanks for the find. I didn't notice the bug in `Clean()` (now `SlashedPathClean()`). I've fixed the bug there and added test cases for root directory. Please resolve this thread if appropriate.

File: internal/fsutil/path_test.go
Comment (woky) [reply internal/fsutil/path_test.go]: Thanks! Indeed it didn't handle `/` correctly. It should be fixed now and there test cases testing various forms of the root directory. Please check it out and resolve if appropriate.

File: internal/fsutil/path.go
Comment (woky) [reply internal/fsutil/path.go:1]: Thanks for providing suggestion with sample code! I've applied it verbatim.

**Comment by woky** [comment]
This PR depends on #69.

**Review by flotter — APPROVED** [review]
Looks great to me!

**Comment by letFunny** [comment]
Per our offline discussion, I think our best way forward is to close this PR and discuss with @rebornplusplus to see if it is a priority right now, and how to amend the code here to make it compatible with the latest changes. We will take that discussion offline and we can always re-open the PRs or create new ones as we see fit.

**Pull request closed without merging**
